primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,404 @@
+// ! 		log::error!("Rpc call error: {:?}", e);
+// ! 	}
+// ! }
+
+use crate::{RuntimeDebug, offchain::Duration, offchain::Timestamp};
+use serde_json::{json, Value};
+use sp_std::{vec::Vec, str};
+use super::http;
+
+/// A rpc call error
+#[derive(Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum Error {
+	/// Response can not be deserialized
+	Deserializing,
+	/// Http error
+	Http(http::Error),
+}
+
+/// A rpc result type
+pub type RpcResult = Result<Response, Error>;
+
+/// A rpc call error field
+#[derive(RuntimeDebug, PartialEq)]
+pub struct RpcError {
+	/// error code
+	pub code: i32,
+	/// error message
+	pub message: Vec<u8>,
+	/// error data
+	pub data: Option<Value>,
+}
+
+/// A rpc call response
+#[derive(RuntimeDebug, PartialEq)]
+pub struct Response {
+	/// rpc response jsonrpc
+	pub jsonrpc: Vec<u8>,
+	/// rpc response result
+	pub result: Option<Value>,
+	/// rpc response error struct
+	pub error: Option<RpcError>,
+	/// rpc response id
+	pub id: u32,
+}
+
+/// A rpc call request
+#[derive(Clone, RuntimeDebug)]
+pub struct Request<'a> {
+	/// rpc request jsonrpc
+	pub jsonrpc: &'a str,
+	/// rpc request id
+	pub id: u32,
+	/// rpc request method
+	pub method: &'a str,
+	/// rpc request params
+	pub params: Vec<&'a str>,
+	/// rpc request timeout
+	pub timeout: u64,
+	/// rpc request url
+	pub url: &'a str,
+}
+
+impl<'a> Default for Request<'a> {
+	fn default() -> Self {
+		Request {
+			jsonrpc: "2.0",
+			id: 1,
+			method: "chain_getFinalizedHead",
+			params: Vec::new(),
+			timeout: 3_000,
+			url: "http://localhost:9933"
+		}
+	}
+}
+
+impl<'a> Request<'a> {
+	/// Create a new Request
+	pub fn new() -> Self {
+		Request::default()
+	}
+
+	/// Change the URL for the Request
+	pub fn url(mut self, url: &'a str) -> Self {
+		self.url = url;
+		self
+	}
+
+	/// Change the Method for the Request
+	pub fn method(mut self, method: &'a str) -> Self {
+		self.method = method;
+		self
+	}
+
+	/// Change the Timeout for the Request
+	pub fn timeout(mut self, timeout: u64) -> Self {
+		self.timeout = timeout;
+		self
+	}
+
+	/// Change the Params for the Request
+	pub fn params(mut self, params: Vec<&'a str>) -> Self {
+		self.params = params;
+		self
+	}
+
+	/// Change the Jsonrpc for the Request
+	pub fn jsonrpc(mut self, jsonrpc: &'a str) -> Self {
+		self.jsonrpc = jsonrpc;
+		self
+	}
+
+	/// Change the ID for the Request
+	pub fn id(mut self, id: u32) -> Self {
+		self.id = id;
+		self
+	}
+
+	/// Send the request and return a RPC result.
+	///
+	/// Err is returned in case there is a Http
+	/// or deserializing error.
+	pub fn send(&self) -> RpcResult {
+		let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(self.timeout));
+		self.send_with_deadline(deadline)
+	}
+
+	/// Send the request and return a RPC result.
+	/// Deadline is used instead of Timeout
+	/// Err is returned in case there is a Http
+	/// or deserializing error.
+	pub fn send_with_deadline(&self, deadline: Timestamp) -> RpcResult {
```

Comment:
That's quite misleading method, cause the `timeout` parameter is now silently ignored.

Reply:
Yes, it is, but I was following your suggestion in the previous comment to allow both options, sending a request with `timeout` or directly defining `deadline`.

> 
```rust
pub fn send(&self, request: &Request) -> ...{
    let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(self.request_timeout));
    self.send_with_deadline(request, deadline);
}
  
pub fn send_with_deadline(&self, request: &Request, deadline: Timestamp) -> ... {
    /// rest of the sending code
}
```

Is it so bad that `timeout` is ignored when using `deadline`? Maybe we can simplify it by making it possible to use only one of them, probably `deadline`.

I didn't remove the method yet, waiting for new suggestions about how you'd like to implement it.